Issue #265 - user panel control

Naltharial opened the pull request:
#265: Added basic support for users to select which panels they should show. Dependencies checked on form validation.

TODO: Required/recommended extension for inter-panel dependencies.

Commit:
Basic start for issue #265 - added form to enable/disable homepage panels.

Commit:
Upstream merge.

Commit:
Panel ordering.

mitar commented on piplmesh/panels/dummy_test/panel.py:
```diff
@@ -0,0 +1,22 @@
+import random
```
Why additional panel? You could just in original panel file extend another panel class and register it. To reduce duplication. So, use inheritance.

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,34 @@ def dispatch(self, request, *args, **kwargs):
+        user = self.request.user
+        user.user_panels.panels_disabled = [key for key, val in form.cleaned_data.iteritems() if not val]
+        user.save()
+        messages.success(self.request, _("You have successfully set your panel defaults."))
```
No need for the message because you are returning to the same page. You would need this if you would be redirecting to some other page and would need to give the user feedback.

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,34 @@ def dispatch(self, request, *args, **kwargs):
+        copying request.POST so we can change users' selection on validation.
+        """
+        kwargs = {'initial': self.get_initial()}
+        if self.request.method in ('POST', 'PUT'):
```
Ehm, forms have only POST.

mitar replied:
This is not REST API.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +165,59 @@ def clean_confirmation_token(self):
         return confirmation_token
+
+def panel_form_factory():
+    """
```
Interesting approach, but it is a bit hackish. ;-) I much more prefer metaclass approach. Here is [done for models](https://github.com/wlanslovenija/mainpage/blob/master/mainpage/wlansi/accounting/models.py#L13), but it is very similar. So you define a function which adds class attributes at creation time. Can you convert your code into it?

mitar commented on piplmesh/account/models.py:
```diff
@@ -48,6 +48,16 @@ class TwitterAccessToken(mongoengine.EmbeddedDocument):
     key = mongoengine.StringField(max_length=150)
     secret = mongoengine.StringField(max_length=150)
 
+class UserPanels(mongoengine.EmbeddedDocument):
```
Can you make model like it is defined [here](https://github.com/wlanslovenija/PiplMesh/issues/278).

So the idea is that you have a list/map of enabled plugins. So anything which is in user object is enabled. Not that you have a flag. There is no reason to have a flag. User adds a panel or removes it. This is it.

mitar commented on piplmesh/account/templates/form/panels.html:
```diff
@@ -0,0 +1,39 @@
+                      	{% trans "Required panels" %}:
+                      	<ul>
+                          {% for dep in field.field.dependencies %}
+                          	<li>{{ dep }}</li>
```
Indentation.

Commit:
Remade form model with metaclass, changed Panels model more similiar to Issue #278.

Naltharial commented:
Changed the form to metaclass, changed the `Panels` model. Not sure why Issue #278 has a double-nested `MapField`, however. Or why we're passing `number_of_columns` as an argument. So right now, each user has a list of their panels and every panel can be collapsed and is at exactly one `(column,order)` position.

Commit:
Remove debugging.

Commit:
Merge remote-tracking branch 'upstream/master'

mitar commented:
Based on the width of the window, number of columns for panels change. So we have double nested mapfiled because we are storing layout of panels for each number of columns there is. So once we map panel id to Panel. And then we map number of columns to PanelState, which tells where for this number of columns panel is.

mitar commented:
So you will have to fix this. :-)

Commit:
Improved consistency with Issue #278.

Commit:
Merge remote-tracking branch 'upstream/master'

Naltharial commented:
Added the second `MapField`. I've just pulled `collapsed` out of `PanelLayout`, because it's not connected to number of columns.

mitar commented:
Not true. It can be. If you have very narrow window on your laptop, you might not want to display all panels. You do not want them to be removed, but just collapsed. On your big monitor on your PC, you have more columns and you want all of panels to be shown fully.

Naltharial commented:
Well, sure, it just wasn't connected currently the way the JavaScript part is set up (requests to `panel/collapse/` do not include `number_of_columns`). I can connect it, but that would mean changing the JavaScript behind it, which I thought was part of another issue.

mitar commented:
OK. You decide. So I am explaining why model was proposed in the way it was. It is true that current code does not behave like that. But the end it should. So you can fix it yourself, or add TODO/ticket for this.

Commit:
Added number_of_columns support to /panels/collapse/

Naltharial commented:
Added the number to the requests for `collapse` and changed the model.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +165,58 @@ def clean_confirmation_token(self):
+    """
+    Form for selecting homepage panels.
+    """
+    __metaclass__ = PanelFormMetaclass
```
Empty line after docstring.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +165,58 @@ def clean_confirmation_token(self):
+        # Add information about dependencies to display in template
+        for name in self.fields:
+            panel = panels.panels_pool.get_panel(name)
+            self.fields[name].dependencies = panel.get_dependencies()
```
I must say that I find this ugly. Attaching some external information to the field. This is OK if there is no other way, but in this case there is: you can retrieve this information ad-hoc in `clean` method. And in template you could simply provide additional data, like dict, which would map between field name and dependencies.

But OK, let it be like this for now. Let see other things.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +165,58 @@ def clean_confirmation_token(self):
+
+class PanelFormMetaclass(forms.Form.__metaclass__):
+    def __new__(cls, name, bases, attrs):
+        panel_list = panels.panels_pool.get_all_panels()
```
Probably no real need for a variable here. Could be simply in the for loop. (Just hint, no need to change, but you can.)

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +165,58 @@ def clean_confirmation_token(self):
+        super(forms.Form, self).__init__(*args, **kwargs)
+        
+        # Add information about dependencies to display in template
+        for name in self.fields:
```
`for name, field in self.fields.items():` does not work? Then you would not have to access dict again bellow. And it might be more readable code.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +165,58 @@ def clean_confirmation_token(self):
+                errors = []
+                for dep in panel.get_dependencies():
+                    if not cleaned_data[dep]:
+                        errors.append(_("Panel '%s' depends on panel '%s', it has been selected." % (name, dep)))
```
OK. This is misuse of the API. Here you should or fix dependencies or display an error. You cannot do both.

mitar replied:
Hm, but it works nicely. :-) OK.

mitar replied:
But, hm. It is not clear to the user if he has to submit form again or not, when dependent panels are automatically selected. With current code, because you are misusing errors, it is not saved until she submits again.

So I see two options:
- you display and error but don't fix the value automatically
- you fix the value automatically but use Django messages framework to display an error (or warning), not form errors

For later you will have to have access to request, which you can pass as a constructor to the form (see `UserCurrentPasswordForm` for example).

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +165,58 @@ def clean_confirmation_token(self):
+        cleaned_data = super(forms.Form, self).clean()
+        add = []
+        
+        while True:
```
I would not display errors for missing dependencies for automatically selected dependencies under the later, but they should be listed under initial panel. So if panel A requires panel B which in turn requires panel C, error message should be "Panel A depends on panels B and C, which have been selected."

mitar replied:
(And not B nor C have been selected. If one is, then message mentions only the other one.)

mitar commented on piplmesh/account/models.py:
```diff
@@ -83,10 +195,17 @@ class User(auth.User):
-    panels_collapsed = mongoengine.DictField()
-    panels_order = mongoengine.DictField()
+    
+    panels = mongoengine.EmbeddedDocumentField(Panels)
```
Ehm. Why you have one more embedded field in-between? So why not just MapField which maps names to Panel embedded documents? If Panel is in the map, panel is activated, if it is not, it is not activated?

Naltharial replied:
Well, because `collapsed` was initially not bound to `number_of_colums`, just to panels. So later I left it like that in case we need to add properties to panels that are not part of column layout, which seemed likely at some point. So it's just for the future, it would be a lot of work to add panel properties later.

mitar replied:
But schema defined in #278 supports this? You have a `Panel` document, which can have next to `layout` value any other property which might not be connected with the layout?

So why are you having one more level of indirection? Compare proposed schema in #278 with yours. Am I missing something?

Naltharial replied:
Connected with the `Layout`, yes, but it doesn't support another property connected with an individual `Panel`, regardless of layout. I can take it out if we won't need it, but it will be difficult to re-add later, if we decide to add a property to a panel.

mitar replied:
How no? Each user has its own list of panels:

```
class PanelState(mongoengine.EmbeddedDocument):
    collapsed = mongoengine.BooleanField()
    column = mongoengine.IntField()
    order = mongoengine.IntField()

class Panel(mongoengine.EmbeddedDocument):
    layout = mongoengine.MapField(mongoengine.EmbeddedDocumentField(PanelState))
    some_property = mongoengine.StringField()

class User(auth.User):
    panels = mongoengine.MapField(mongoengine.EmbeddedDocumentField(Panel))
```

You can then even have panels define their own custom document:

```
class MyPanel(Panel):
    some_additional_property_just_for_this_type = mongoengine.StringField()
```

And `User`'s `panels` field then maps between ids of panels and those documents. For now we have just default Panel document, so it is easy. But later on each of them can be also custom documents, as defined above, inheriting from the common document.

So, what exactly is not possible to do here and why would your solution be superior to mine?

(Just to be sure, I like that you argue and are not giving up, so please continue and provide arguments. It might be that I am really not seeing something, but I do not like the idea of one more level of indirection if it is not necessary.)

Naltharial replied:
Ah, I see now in this comment which part you were referring to. I think my model is similiar, I just connected the objects a bit differently. But, from your example, I can see what you meant `Panel` to really be. I've renamed the classes to what they are here and changed the access to be more consistent, with less layers.

Probably my PHP habits to create layers because the language is not organized.

mitar commented on piplmesh/account/templates/form/panels.html:
```diff
@@ -0,0 +1,39 @@
+{% load i18n %}
```
OK. This will have to be improved. It is mostly duplication of form.html. This is not good. It should be done in extendable way. But I will do that.

mitar replied:
(After I pull this in.)

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +165,58 @@ def clean_confirmation_token(self):
+        while True:
+            if add:
+                # User has not selected some of the dependencies, so we select them
+                for val in add:
```
Replace `val` with something meaningful. What is meaning of `val`? Value of what? What does it contain?

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,32 @@ def dispatch(self, request, *args, **kwargs):
+
+    def form_valid(self, form):
+        user = self.request.user
+        user.panels.set_panels([key for key, val in form.cleaned_data.iteritems() if val])
```
Replace `val` with something meaningful. What is meaning of `val`? Value of what? What does it contain?

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -123,27 +123,30 @@ def send_update_on_new_post(sender, post, request, bundle, **kwargs):
-        panels = []
-
+        order = 0
+        oldc = None
```
`oldc` is what? Be verbose.

mitar commented on piplmesh/panels/dummy/panel.py:
```diff
@@ -15,6 +15,18 @@ def get_context(self, context):
         })
         return context
+   
+class DummyTestPanel(DummyPanel):
```
`DummyTestPanel` -> `DummyDependencyPanel`

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,32 @@ def dispatch(self, request, *args, **kwargs):
+        kwargs = {'initial': self.get_initial()}
+        if self.request.method == 'POST':
+            kwargs.update({
+                'data': dict(self.request.POST),
```
I don't understand how this helps?

mitar replied:
Hm, the only thing you are doing here is making data mutable?

Naltharial replied:
It's so we can change which panels the user ticked if they didn't select all of the prerequisites. If we're not changing the selection anymore (not using form errors) then it'll probably not be neccessary anymore.

mitar replied:
There is difference between displaying error messages and selecting dependencies automatically. So this is needed to be able to select dependencies automatically. But I am just wondering here why it is necessary for it to be here and not for example to copy this in view or form constructor.

mitar replied:
Anyway, I think we could do this better. Dependencies should be resolved in JavaScript on the client. So user should get feedback there, dynamically.

On the server I would just test if everything is correct and display an error. Error should be possible in practice only if somebody is doing something nasty. In reality JavaScript would take care that everything is correct. But we just check for every case.

So this your approach to require user to submit the form to verify it is so 90'.

Commit:
Moved dependency checking to client side.

Naltharial commented:
I've moved most of the checking to `panels_validate` view, which is called by JavaScript when user input is to be validated, before submitting the form.

mitar commented:
Please update from main repository. It cannot be automatically merged.

mitar commented:
Why? Why are you requiring call to the server? I would agree if validation would be very complex. But here it is simple and is especially important user feedback. So what you do on error. And if you code that in JavaScript, you can provide much nicer feedback there. So simply, you can have a form validation which checks if all dependencies are resolved and raises error if not. And then more complex JavaScript code which can in real-time, as user clicks around, enable dependent panels automatically and write some messages bellow that panels have been enabled because they are dependent.

Commit:
Dependency checking completely client side.

Naltharial commented:
Removed AJAX, sent panels as JSON to client so checking can be done completely in JavaScript.

I've tried merging, but PiplMesh stops working if I do:

```
  File "D:\My Documents\FRI\IPRI\PiplMesh\PiplMesh\piplmesh\frontend\views.py", line 128, in <module>
@mongoengine_signals.post_save.connect_via(sender=api_models.Notification)
AttributeError: '_FakeSignal' object has no attribute 'connect_via'
```

if I remove that signal, the updates are being requested at `http://127.0.0.1:8001/updates/user/[GUID]/`, which pushserver returns 404 on. Should I update anything else except `pip install -r requirements.txt`?

mitar commented on piplmesh/account/templates/user/panels.html:
```diff
@@ -0,0 +1,28 @@
+    <script src="{% static "piplmesh/js/panels.js" %}" type="text/javascript"></script>
+    <script type="text/javascript">
+        /* <![CDATA[ */
+        var panels = $.parseJSON('{{ panels|safe }}');
```
Why is `parseJSON` needed? JSON should be parsed automatically?

mitar replied:
I would rather see that you pass Python object to the template and then in the template use filter which escapes into JSON.

mitar commented:
Please update from main repository, so that it can be automatically merged.

mitar commented:
As you can see [here](https://github.com/MongoEngine/mongoengine/blob/master/mongoengine/signals.py), `_FakeSignal` is used only if you do not have `blinker` package available. Have you installed it with pip? It is in `requirements.txt`.

mitar commented:
> which pushserver returns 404 on.

Are your regex for `url` same as [here](https://github.com/wlanslovenija/PiplMesh/blob/master/piplmesh/settings.py#L203)? So `(.+)`?

Commit:
Upstream merge, json filter.

Commit:
Validation JavaScript.

Commit:
Merge remote-tracking branch 'upstream/master'

Naltharial commented:
I ran requirements through. It looks like the blinker entry didn't work. I got

```
Downloading/unpacking blinker==1.2 (from -r requirements.txt (line 9))
```

and the setup conpleted, but then still the `_FakeSignal` error occured. I had to manually run `pip install blinker` to fix it. Maybe a version issue? The pushserver issue went away after a while, so the merge should be completed now.

I created a JSON filter instead of doing it manually.

mitar commented:
Which version of blinker does `pip freeze` report for you now?

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +166,39 @@ def clean_confirmation_token(self):
+    def __init__(self, *args, **kwargs):
+        super(PanelForm, self).__init__(*args, **kwargs)
+        
+        # Add information about dependencies to display in template
```
Do we need this now? Not anymore?

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +166,39 @@ def clean_confirmation_token(self):
+class PanelFormMetaclass(forms.Form.__metaclass__):
+    def __new__(cls, name, bases, attrs):
+        for panel in panels.panels_pool.get_all_panels():
+            attrs[panel.get_name()] = forms.BooleanField(required=False, initial=False)
```
Do we really have to supply initial value here? Do we not always populate it from state in the database?

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +166,39 @@ def clean_confirmation_token(self):
+            panel = panels.panels_pool.get_panel(panel_name)
+            for dep in panel.get_dependencies():
+                if not cleaned_data[dep]:
+                    raise exceptions.ValidationError(_('Dependencies not satisfied.'))
```
User exposed strings should be in `"`.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +166,39 @@ def clean_confirmation_token(self):
+                continue
+            
+            panel = panels.panels_pool.get_panel(panel_name)
+            for dep in panel.get_dependencies():
```
Be verbose, `dep` -> `dependency`.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +166,39 @@ def clean_confirmation_token(self):
+    def clean(self):
+        cleaned_data = super(PanelForm, self).clean()
+        
+        for panel_name, panel_checked in cleaned_data.items():
```
`panel_checked` -> `panel_enabled`, use better semantic meaning for variable names.

mitar commented on piplmesh/account/templatetags/account.py:
```diff
@@ -10,3 +13,9 @@ def user_image(context, user=None):
         'user_image_url': user.get_image_url(),
     }
+
+@register.filter(name='json', is_safe=True)
```
No, this is not really a correct template tag library. It is not really connected with accounts, no?

We have a better one [here](https://github.com/wlanslovenija/PiplMesh/blob/master/piplmesh/frontend/templatetags/frontend.py).

mitar commented on piplmesh/account/templatetags/account.py:
```diff
@@ -1,4 +1,7 @@
 from django import template
+from django.core.serializers import serialize
+from django.utils import safestring, simplejson
+from django.db.models import query
```
Order alphabetically.

mitar commented on piplmesh/account/templatetags/account.py:
```diff
@@ -10,3 +13,9 @@ def user_image(context, user=None):
+
+@register.filter(name='json', is_safe=True)
+def json(obj):
+    if isinstance(object, query.QuerySet):
```
`object` or `obj`?

mitar replied:
(I prefer verbose names.)

mitar commented on piplmesh/account/templatetags/account.py:
```diff
@@ -10,3 +13,9 @@ def user_image(context, user=None):
+def json(obj):
+    if isinstance(object, query.QuerySet):
+        return safestring.mark_safe(serialize('json', obj))
+    return safestring.mark_safe(simplejson.dumps(obj))
```
This is a good question. Is JSON always safe in HTML way safe? So Django supports only HTML-context automatic escaping. By marking it safe, you are saying that JSON is safe to be included in HTML context as it is. Is this really true? In any calling context?

I think it is better to not assume this and leave to users to use filters in templates to mark this. So when user knows that context is safe, she can use `safe` template filter additionally after your tag.

mitar commented on piplmesh/account/templatetags/account.py:
```diff
@@ -1,4 +1,7 @@
 from django import template
+from django.core.serializers import serialize
```
Import only modules.

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,26 @@ def dispatch(self, request, *args, **kwargs):
+
+    def form_valid(self, form):
+        user = self.request.user
+        user.panels.set_panels([panel for panel, checked in form.cleaned_data.items() if checked])
```
`checked` -> `enabled`.

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,26 @@ def dispatch(self, request, *args, **kwargs):
+    def get_context_data(self, **kwargs):
+        context = super(PanelView, self).get_context_data(**kwargs)
+        
+        context['panels'] = {panel.get_name(): panel.dependencies for panel in self.request.user.panels.get_all_panels()}
```
Wrap.

mitar replied:
Or don't. Forget this one.

mitar commented:
OK. So we have to decide on schema. Because some code things are related to that.

Commit:
Consistent panel model, improvements.

Naltharial commented:
Strangely, it reports the same version, 1.2. Might be something with me using Windows, probably not a general bug.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +166,34 @@ def clean_confirmation_token(self):
+    __metaclass__ = PanelFormMetaclass
+    
+    def __init__(self, *args, **kwargs):
+        super(PanelForm, self).__init__(*args, **kwargs)
```
And then all this is unnecessary. :-)

mitar commented on piplmesh/account/forms.py:
```diff
@@ -164,3 +166,34 @@ def clean_confirmation_token(self):
+            
+            panel = panels.panels_pool.get_panel(panel_name)
+            for dependency in panel.get_dependencies():
+                if not cleaned_data[dependency]:
```
Does not this throw a `KeyError` if dependency is missing? Probably it should be `cleaned_data.get(dependency, False)`.

mitar commented on piplmesh/panels/dummy/panel.py:
```diff
@@ -15,6 +15,18 @@ def get_context(self, context):
+        context = super(DummyDependencyPanel, self).get_context(context)
+
+        context.update({
+            'header': _("Dummy test panel"),
```
`test` -> `dependency`.

mitar commented on piplmesh/panels/base.py:
```diff
@@ -16,8 +16,8 @@ def __init__(self):
-            if not panels_pool.panels_pool.has_panel(dependency):
-                raise exceptions.PanelDependencyNotRegistered("Panel '%s' depends on panel '%s', but later is not registered" % (self.get_name(), dependency))
+            if not panels_pool.has_panel(dependency):
+                raise exceptions.PanelDependencyNotRegistered("Panel '%s' depends on panel '%s', but the latter is not registered" % (self.get_name(), dependency))
```
Vau, I didn't know that `later` is different than `latter`.

mitar commented on piplmesh/panels/base.py:
```diff
@@ -16,8 +16,8 @@ def __init__(self):
         for dependency in self.get_dependencies():
-            if not panels_pool.panels_pool.has_panel(dependency):
-                raise exceptions.PanelDependencyNotRegistered("Panel '%s' depends on panel '%s', but later is not registered" % (self.get_name(), dependency))
+            if not panels_pool.has_panel(dependency):
```
Are you sure it should be `panels_pool.has_panel` and not `panels_pool.panels_pool.has_panel`? This is quite tricky, but here we import first `panels_pool` module, which has `panels_pool` attribute. I think it should be `panels_pool.panels_pool`. Have you tested this?

Naltharial replied:
I get

```
'PanelsPool' object has no property 'panels_pool'
```

if I try it.

mitar commented on piplmesh/frontend/templatetags/frontend.py:
```diff
@@ -1,7 +1,23 @@
+@register.filter
+def lookup(dict, key):
+    if key in dict:
+        return dict[key]
```
Hm, return `dict.get(key)`? Maybe just rename `lookup` to `get`? :-)

mitar commented on piplmesh/frontend/templatetags/frontend.py:
```diff
@@ -1,7 +1,23 @@
 def is_active(current_path, url_path):
     return current_path.startswith(url_path)
+
+@register.filter(name='json', is_safe=True)
```
No need for `name=` parameter, by default it is the same as name of the function.

mitar replied:
I am not sure that `is_safe=True` is correct here, too.

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,26 @@ def dispatch(self, request, *args, **kwargs):
+        return super(PanelView, self).form_valid(form)
+    
+    def get_initial(self):
+        return dict(zip(self.request.user.panels.keys(), [True] * len(self.request.user.panels)))
```
Beautiful. If you want obfuscation. :-)

Is maybe this better?

```
{panel: True for panel in self.request.user.panels.keys()}
```

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,26 @@ def dispatch(self, request, *args, **kwargs):
+    def get_context_data(self, **kwargs):
+        context = super(PanelView, self).get_context_data(**kwargs)
+        
+        context['panels'] = {panel.get_name(): panel.dependencies for panel in self.request.user.get_all_panels()}
```
Use `context.update` structure, as it is easier to extend later on.

Rename `panels` to something else. `dependencies`? Something which tells what the content is.

mitar replied:
So you have three levels. `panels_dependencies` -> `panel_dependencies` -> `panel_dependency`. Ugly. Find something better. :-)

Naltharial replied:
Hmm, it's a dict where the key is panel name and the value a list of dependencies. I'm not sure how to do it better than one dict?

```
context.update({'dependencies': {panel.get_name(): panel.dependencies for panel in self.request.user.get_all_panels()}})
```

mitar replied:
Sorry. I was not clear enough. Structure is perfect here. So dict is a good choice. I had problems only with name. So `panels` name is not good enough. and later on in the code you need three levels of elements (it is OK to need three level of elements), so how would we name the top element, so that lower elements would also have a good name. So if you start here with `dependencies`, how you iterate over its nested elements in a way that it is understandable. I proposed `panels_dependencies` -> `panel_dependencies` -> `panel_dependency`, but I find it ugly.

(The other minor thing was to use `context.update`.)

Naltharial replied:
`panels_x_dependencies` -> `dependencies` -> `dependency`?

The first is a relational database construct used for many-to-many relations, which this is, the second two are just dependency because dependencies are always of a panel. I don't know, it's kind of mixing realms of programming here, but it's a many-to many relation technically. Not sure how much better you can get with just names. :)

Other than `_x_`, some use `_to_`, `_per_`, `_has_`, or just `X`.

mitar replied:
He he. I don't know either. :-) I don't like X.

But you decide. My main comment was that you should find something and that it should be consistent.

mitar commented on piplmesh/account/templates/user/panels.html:
```diff
@@ -0,0 +1,24 @@
+    <script src="{% static "piplmesh/js/panels.js" %}" type="text/javascript"></script>
+    <script type="text/javascript">
+        /* <![CDATA[ */
+        var panels = $.parseJSON('{{ panels|json|safe }}');
```
So `var panels = {{ panels|json|safe }}` does not work?

Otherwise, you can try:

```
var panels = $.parseJSON('{{ panels|json|escapejs }}');
```

mitar commented on piplmesh/account/templates/user/panels.html:
```diff
@@ -0,0 +1,24 @@
+{% extends "plain.html" %}
+
+{% load future i18n staticfiles frontend %}
```
`future` is not needed because we are not using `url` tag anywhere.

mitar commented on piplmesh/account/templates/form/panels.html:
```diff
@@ -0,0 +1,41 @@
+                            {{ field.label }}
+                        </label>
+                        {% with panels|lookup:field.name as panel_dependencies %}
+                        {% if panel_dependencies %}
```
Indent.

mitar commented on piplmesh/account/templates/form/panels.html:
```diff
@@ -0,0 +1,41 @@
+                        {% with panels|lookup:field.name as panel_dependencies %}
+                        {% if panel_dependencies %}
+                            <div class="dependencies">
+                                {% trans "Required panels" %}:
```
No, `:` should be in translation string. Not all languages use `:` the same. (French for example have space before it.)

mitar commented on piplmesh/account/templates/form/panels.html:
```diff
@@ -0,0 +1,41 @@
+{% load i18n frontend %}
+
+<form action="{{ form_action }}" method="post" id="panel_form" onsubmit="return validatePanels(this)">
```
No, attach events with jQuery. Not in template/HTML.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,58 @@
```
Indentation. Use four spaces.

mitar replied:
JavaScript indentation is not good in the file.

mitar replied:
(Maybe there are tabs?)

mitar commented on piplmesh/account/models.py:
```diff
@@ -53,6 +53,20 @@ class TwitterAccessToken(mongoengine.EmbeddedDocument):
+    
+    def get_layout(self, columns):
+        if columns in self.layout:
+            return self.layout[columns]
```
Do you maybe have whole this function just because you do not know that you could do `self.layout.get(columns)`?

mitar commented on piplmesh/account/models.py:
```diff
@@ -53,6 +53,20 @@ class TwitterAccessToken(mongoengine.EmbeddedDocument):
+        if columns in self.layout:
+            return self.layout[columns]
+        else:
+            return False
```
In general it is better to return `None` in such cases.

mitar commented on piplmesh/account/models.py:
```diff
@@ -89,10 +103,16 @@ class User(auth.User):
+        # If the user has not previously saved any panel data, set up defaults
+        if not self.panels:
+            self.reset_panels()
+            self.save
```
This is not good. You do not really want document to be automatically saved on initialization. So normal MongoEngine way is to do:

```
object = Document()
object.foo = 123
# nothing is saved yet
object.save()
```

In your case, document is automatically saved on initialization.

mitar commented on piplmesh/account/models.py:
```diff
@@ -89,10 +103,16 @@ class User(auth.User):
+    def __init__(self, *args, **kwargs):
+        super(User, self).__init__(*args, **kwargs)
+        
+        # If the user has not previously saved any panel data, set up defaults
```
Why `default` parameter to `mongoengine.MapField` above is not good enough for this purpose?

mitar replied:
There is another downside of this. This means that user will automatically have enabled only those panels which existed when user object was initially created. If later on we add new panel, all users will have to manually add those panels. Is there anything we can do about it?

Naltharial replied:
TODO as of  wlanslovenija/PiplMesh@6a06c735e07f4d881d28d43f6405dd335fdc4da3

mitar replied:
Don't understand your comment. And I also see that mine is on the strange line.

mitar commented on piplmesh/account/models.py:
```diff
@@ -53,6 +53,20 @@ class TwitterAccessToken(mongoengine.EmbeddedDocument):
+    order = mongoengine.IntField()
+    
+class Panel(mongoengine.EmbeddedDocument):
+    layout = mongoengine.MapField(mongoengine.EmbeddedDocumentField(PanelState))
```
Probably a comment about this field would be useful, to tell that we are mapping between number of columns to PanelState. Do we store number of columns as number or as string?

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -162,27 +163,31 @@ def test_if_running_as_celery_worker():
-
+        order = 0
+        old_column = None
+        columns, orders = dict(), dict()
```
`dict()` => `{}`

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -162,27 +163,31 @@ def test_if_running_as_celery_worker():
+            columns[name] = column
+            orders[name] = order
+        
+        request.user.set_panels(request.POST.getlist('names'), request.POST['number_of_columns'], column=columns, order=orders)
```
Don't reuse same function for two different function calls. Define another function.

mitar replied:
Why don't you simply have two arguments, number of columns, and dict of names mapping to tuple of (column, order)? Now you have names three times, in names, columns keys, orders keys. Error prone if by accident they do not match. It is not smart to define API (even internal one) which requires three different parameters to match like that.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,58 @@
+function validatePanels(form) {
+	var status = true;
+	var selected_panels = {}
```
`;` at the end.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,58 @@
+	var status = true;
+	var selected_panels = {}
+	$('[type=checkbox]', $(form)).each(function() {
+		if ($(this).prop('checked')) {
```
I think that `.is(':checked')` is the proposed way to see if checkbox is checked?

Naltharial replied:
Well, as per http://api.jquery.com/prop/ both are suggested, but I think `prop` is accessed though DOM and `is` is though CSS, so I personally prefer `prop`. This is only jQ 1.6+ I think.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,58 @@
+function validatePanels(form) {
+	var status = true;
+	var selected_panels = {}
+	$('[type=checkbox]', $(form)).each(function() {
```
In anonymous functions, put space before `()`. Write out all arguments of the callback, even if you are not using it. So later developers will know what is available.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,58 @@
+		}
+	});
+	
+	for (var panel in selected_panels) {
```
Use jQuery `.each` to iterate over, then you do not need to play with `hasOwnProperty`.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,58 @@
+			}
+		}
+		
+		alert(error);
```
Oh, this is not really user friendly. I imagine rather that instead of saying what is wrong and then user has to manually fix it, is much better to prevent user to get to this state in the first place.

So I would rather do it that when you enable some panel which has dependencies, all dependent panels' checkboxes become checked, but disabled for editing. And bellow it should be written something like "Automatically enabled because panel1 and panel2 dependency."

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
         return user
+    
+    def get_layouts(self, columns):
+        return dict([(name, panel.get_layout(columns) if panel.get_layout(columns) else PanelState()) for name, panel in self.panels.items()])
```
Ehm, `panel.get_layout(columns) or PanelState()` is probably nicer?

mitar replied:
And again, `{name: panel.get_layout(columns) or PanelState() for name, panel in self.panels.items()}`.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+        return panels.panels_pool.get_all_panels()
+    
+    def get_collapsed(self, columns):
+        collapsed = [(name, panel.get_layout(columns).collapsed if panel.get_layout(columns) else False) for name, panel in self.panels.items()]
```
Sometimes think about reusing existing other functions. For example, here you could call first `get_layouts` and then process its outputs.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+               column = self.panels[name].layout[columns].column if self.panels[name].get_layout(columns) else None,
+               order = self.panels[name].layout[columns].order if self.panels[name].get_layout(columns) else None,
+            )
+            
```
Unnecessary empty line.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+        collapsed = [(name, panel.get_layout(columns).collapsed if panel.get_layout(columns) else False) for name, panel in self.panels.items()]
+        return dict(collapsed)
+    
+    def set_collapsed(self, columns, name, collapsed):
```
`columns`!!! Decide how you name this. What is `columns`? List of columns? Or number of columns? Name it the same in caller and callee.

mitar replied:
Maybe we should always everywhere name it `columns_count`?

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+        return dict(collapsed)
+    
+    def set_collapsed(self, columns, name, collapsed):
+        self.panels[name].layout[columns] = PanelState(
```
Are you sure that `.layout` is always defined?

mitar replied:
Forget that. It should be.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+    
+    def set_collapsed(self, columns, name, collapsed):
+        self.panels[name].layout[columns] = PanelState(
+               collapsed = collapsed,
```
In function calls we do not have spaces around `=`.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+        self.panels[name].layout[columns] = PanelState(
+               collapsed = collapsed,
+               column = self.panels[name].layout[columns].column if self.panels[name].get_layout(columns) else None,
+               order = self.panels[name].layout[columns].order if self.panels[name].get_layout(columns) else None,
```
This two lines are ugly.

Why not have everything something like:

```
layout = self.panels[name].get_layout(columns) or PanelState()
layout.collapsed = collapsed
self.panels[name].layout[columns] = layout
```

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+                   )
+    
+    def reset_panels(self):
+        for panel in panels.panels_pool.get_all_panels():
```
Use `self.get_all_panels()` here.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+    
+    def reset_panels(self):
+        for panel in panels.panels_pool.get_all_panels():
+            self.panels[panel.get_name()] = Panel()
```
And probably you want to do rather:

```
self.panels = {panel.get_name(): Panel() for panel in self.get_all_panels()}
```

mitar replied:
Your code does not reset values for old panels which are not anymore in `get_all_panels`.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+        # Preserve prior settings for kept panels
+        self.panels = dict((name, self.panels[name] if name in self.panels else Panel()) for name in panels)
+        
+        # If number of columns was passed as the first argument, we're also reordering a layout
```
Split this into two functions.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+                # If properties were passed as dicts in kwargs use them, otherwise use existing or default
+                self.panels[panel].layout[args[0]] = PanelState(
+                   collapsed = kwargs['collapsed'][panel] if 'collapsed' in kwargs else
+                       self.panels[panel].layout[args[0]].collapsed if args[0] in self.panels[panel].layout else False,
```
Ehm, this is really the simplest way to do it?

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+    
+    def get_columns(self, columns):
+        ordered = False
+        cols = dict()
```
See, here you are then mixing `cols` and `columns`, where first are really columns, but the second is just number.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+        cols = dict()
+        cols_order = dict()
+        unordered = []
+        for key, panel in self.get_layouts(columns).items():
```
`key` -> `name`?

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+                cols[c] = []
+                cols_order[c] = []
+
+            # Enforce ordering of panels in column, because layout is not ordered
```
Hm, you probably want to sort/order the dict here? There are easier ways to do it. Read [here](http://wiki.python.org/moin/HowTo/Sorting/) and in general Django has a ordered dict structure available (so it keeps the order). Or simply just convert it to a list and sort.

Naltharial replied:
TODO as of  wlanslovenija/PiplMesh@6a06c735e07f4d881d28d43f6405dd335fdc4da3

mitar replied:
Don't understand your comment. And I also see that mine is on the strange line.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +203,79 @@ def create_user(cls, username, email=None, password=None):
+                cols[len(cols)] = unordered
+            return cols
+        else:
+            return ''
```
`''`?

mitar commented:
So overall things are looking good. But there are some slight improvements still needed. In the internal API you defined. And in user interface.

If you are in hurry, you can maybe open ticket for user interface improvements.

Commit:
Quality improvements, panel form dependency feedback.

Commit:
Upstream merge.

mitar commented:
Tests are failing. See [details](http://travis-ci.org/wlanslovenija/PiplMesh/builds/2603413).

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,25 @@ def dispatch(self, request, *args, **kwargs):
+    
+    def get_context_data(self, **kwargs):
+        context = super(PanelView, self).get_context_data(**kwargs)
+        context.update({'panels_with_dependencies': {panel.get_name(): panel.dependencies for panel in self.request.user.get_all_panels()}})
```
Wrap this into multiple lines.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -193,27 +193,30 @@ def test_if_running_as_celery_worker():
 def panels_collapse(request):
     if request.method == 'POST':
-        request.user.panels_collapsed[request.POST['name']] = True if request.POST['collapsed'] == 'true' else False
+        request.user.set_collapsed(request.POST['number_of_columns'], request.POST['name'], True if request.POST['collapsed'] == 'true' else False)
```
Ha ha. I just now noticed. `True if request.POST['collapsed'] == 'true' else False` is same as `request.POST['collapsed'] == 'true'`.

mitar commented on piplmesh/account/models.py:
```diff
@@ -89,10 +107,15 @@ class User(auth.User):
-    panels_collapsed = mongoengine.DictField()
-    panels_order = mongoengine.DictField()
+    
+    panels = mongoengine.MapField(mongoengine.EmbeddedDocumentField(Panel))
```
You could just define default value here.

Naltharial replied:
How do you define default value for a MapField? Any way I try to pass in a dictionary, it just returns a `ValidationError` of

```
Only dictionaries may be used in a DictField: ['panels']
```

mitar replied:
What exactly are you passing in?

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +206,72 @@ def create_user(cls, username, email=None, password=None):
+        
+        for panel, layout in self.get_layouts(columns_count).items():
+            layout.column = column_ordering[panel][0]
+            layout.order = column_ordering[panel][1]
```
This could be better written as:

```
layout.column, layout.order = column_ordering[panel]
```

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +206,72 @@ def create_user(cls, username, email=None, password=None):
+        """
+        This method reorders panels for a user.
+        
+        columns_count(int): count of columns for which the layout is changed
```
BTW, there is semi-standard way to define inputs and such in docstrings, but OK.

Naltharial replied:
Hmm, I tried looking, but I don't think I found any (well, `backends.py` lists fields, but not types). Is there an example for docstrings?

mitar replied:
Read about [info field lists](http://sphinx.pocoo.org/domains.html#info-field-lists).

Naltharial replied:
So

```
:param dict layout: mapping of count of displayed columns(string) to PanelState() objects for that count
```

?

mitar replied:
```
:param dict layout: mapping of count of displayed columns (string) to PanelState() objects for that count
```

mitar replied:
Or (I don't remember if string should be in one backticks or two, sometimes you can make it link to Python documentation then):

```
:param dict layout: mapping of count of displayed columns(`string`) to PanelState() objects for that count
```

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +206,72 @@ def create_user(cls, username, email=None, password=None):
+        
+        if ordered:
+            if unordered_panels:
+                columns[len(columns)] = unordered_panels
```
`columns[len(columns)]` == `columns[-1]`

mitar commented on piplmesh/account/templates/user/panels.html:
```diff
@@ -0,0 +1,24 @@
+    <script src="{% static "piplmesh/js/panels.js" %}" type="text/javascript"></script>
+    <script type="text/javascript">
+        /* <![CDATA[ */
+        var panels_with_dependencies = {{ panels_with_dependencies|json|safe }}
```
`;` missing at the end.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -193,27 +193,30 @@ def test_if_running_as_celery_worker():
-        panels = request.user.panels_order.get(number_of_columns, [])
-        return http.HttpResponse(simplejson.dumps(panels), mimetype='application/json')
+        return http.HttpResponse(simplejson.dumps(request.user.get_columns(number_of_columns)), mimetype='application/json')
+
```
Unnecessary line.

mitar commented on piplmesh/account/templates/form/panels.html:
```diff
@@ -0,0 +1,43 @@
```
Add a template comment with TODO that this file should be made in a way which extends default `form.html` template.

So, this, that you copy-pasted the whole file and just changed a bit, this is not a good idea in general. There are solutions to it, but we will not go into this now.

mitar replied:
Please update from main repository. I have updated form template there to allow extending it. You can now instead of manually copy-pasting the original template and changing things, rather extend that one and just inject necessary data where you want.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+var panels_forcedby = {}
```
Missing `;` at the end.

mitar replied:
It is not a good idea to define global variables. JavaScript shares all global variables all around. A good practice is to keep them really to minimum or really just to those you have to use to communicate between different scripts.

I would put both those variables and function definitions at the end into `$(document).ready({ ... })` scope.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+		
+		// Trigger event to recursively check dependencies throughout hierarchy.
+		panel_checkbox.change()
+	} else {
```
We go into newline after `}`.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+	if (typeof panels_forcedby[panel] != "undefined" && panels_forcedby[panel].length) {
+		var text = gettext('Panel depended on by:');
+		panel_locktext.text(text);
+		$('<ul><li>' + panels_forcedby[panel].join("\n\t") + '</li></ul>').appendTo(panel_locktext);
```
Generate DOM fragment using jQuery and build it together, don't blindly concatenate things together. Who will escape things in such case?

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+	var panel_checkbox = $('#panel_form input[name="'+panel+'"]');
+	var panel_locktext = $('.locked', panel_checkbox.parents('li'));
+	
+	if (typeof panels_forcedby[panel] != "undefined" && panels_forcedby[panel].length) {
```
For constant stings, use `'`.

mitar replied:
You could use [isEmptyObject](http://api.jquery.com/jQuery.isEmptyObject/) then here.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+var panels_savestate = {}
+
+$(document).ready(function () {
+	var inputs = $('#panel_form input[type="checkbox"]');
```
It is common in jQuery to chain things together. So you do not really have to define a variable here, but just chain all following calls together.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+
+$(document).ready(function () {
+	var inputs = $('#panel_form input[type="checkbox"]');
+	inputs.on('change', function () {
```
Write out all arguments callback functions receive. In this case, `event`.

mitar replied:
(This is useful for future developers to know what is available.)

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+    	validateCheckbox(this);
+    });
+    
+	inputs.each(function (index, element) {
```
Name argument better. We know it is an element. But it is anything more maybe? An input?

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+$(document).ready(function () {
+	var inputs = $('#panel_form input[type="checkbox"]');
+	inputs.on('change', function () {
+    	validateCheckbox(this);
```
BTW, do we really need this short wrapper function? Why if you would simply give `validateCheckbox` as a callback?

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+    });
+    
+	inputs.each(function (index, element) {
+    	var checkbox = $(element);
```
All this could be simplified: `$(':checked', element).change()`.

mitar replied:
(Untested code. But to give you an idea.)

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+    });
+	
+	$('#panel_form').on('submit', function () {
+		$('#panel_form input[type="checkbox"][disabled]').each(function (index, element) {
```
`$('#panel_form input[type="checkbox"][disabled]')` -> `$('input[type="checkbox"][disabled]', this)`

mitar replied:
No need for `.each`. Just:

```
$('input[type="checkbox"][disabled]', this).prop('disabled', false);
```

You think too much in a procedural way. :-) jQuery is quite functional. You can chain, do things over many thing at the same time, and so on.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+		panel_checkbox.change()
+	} else {
+		if (panel_checkbox.prop('disabled')) {
+			panel_checkbox.prop('disabled', false);
```
You can chain those together. Even if the check in if could be probably made into a selector case which would select only disabled subset of `panel_checkbox`, making set empty or leave it at `panel_checkbox`.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+    }
+}
+
+
```
Unnecessary empty line.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+    var panel = checkbox.attr('name');
+    var checked = checkbox.prop('checked');
+    
+    if (typeof panels_with_dependencies[panel] != "undefined" && panels_with_dependencies[panel].length) {
```
I think that `panel in panels_with_dependencies` is nicer than `typeof panels_with_dependencies[panel] != "undefined"`.

mitar replied:
Of course, the question is if it is really needed. This might be enough:

```
$.each(panels_with_dependencies[panel] || [], ...
```

Or does `$.each` even throw an exception, if the thing you pass it in is undefined?

mitar replied:
Or you can do also:

```
panels_with_dependencies[panel] ||= []
```

before.

mitar replied:
(I am not sure if this works, though.)

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+    var checked = checkbox.prop('checked');
+    
+    if (typeof panels_with_dependencies[panel] != "undefined" && panels_with_dependencies[panel].length) {
+	    $.each(panels_with_dependencies[panel], function (key, dependency) {
```
`key` -> name it better. `panel` is a good candidate. :-) `panel_name` maybe, too.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+	    		panels_forcedby[dependency] = [];
+	    	}
+	    	
+	    	var index = panels_forcedby[dependency].indexOf(panel);
```
Use jQuery `inArray`. Does it die id `panels_forcedby[dependency]` would not exist?

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+    if (typeof panels_with_dependencies[panel] != "undefined" && panels_with_dependencies[panel].length) {
+	    $.each(panels_with_dependencies[panel], function (key, dependency) {
+	    	if (typeof panels_forcedby[dependency] == "undefined") {
+	    		panels_forcedby[dependency] = [];
```
```
panels_forcedby[dependency] ||= [];
```

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+	    	}
+	    	
+	    	var index = panels_forcedby[dependency].indexOf(panel);
+	    	if (checked && index < 0) {
```
Aaa, you are using this as a set. Why then don't use set? That is, object?

```
panels_forcedby[dependency][panel] = checked
```

?

mitar replied:
OK. Or:

```
if (checked) {
    panels_forcedby[dependency][panel] = True;
}
elif (!checked && panel in panels_forcedby[dependency]) {
    delete panels_forcedby[dependency][panel];
}
```

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+	var panel_locktext = $('.locked', panel_checkbox.parents('li'));
+	
+	if (typeof panels_forcedby[panel] != "undefined" && panels_forcedby[panel].length) {
+		var text = gettext('Panel depended on by:');
```
No need for temporary variable here?

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+		
+		panels_savestate[panel] = panel_checkbox.prop('checked');
+		panel_checkbox.prop('checked', true);
+		panel_checkbox.prop('disabled', true);
```
Chain those together.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+
+function lockCheckbox(panel) {
+	var panel_checkbox = $('#panel_form input[name="'+panel+'"]');
+	var panel_locktext = $('.locked', panel_checkbox.parents('li'));
```
Oh, but you do know about this syntax.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,72 @@
+
+
+function lockCheckbox(panel) {
+	var panel_checkbox = $('#panel_form input[name="'+panel+'"]');
```
Spaces around `+`.

mitar commented:
Tests are failing. Run `./manage.py test` and see.

mitar commented:
Please update from main repository.

Commit:
Improved panels control.

Commit:
Upstream merge.

mitar commented on piplmesh/account/views.py:
```diff
@@ -402,6 +402,27 @@ def dispatch(self, request, *args, **kwargs):
+    
+    def get_context_data(self, **kwargs):
+        context = super(PanelView, self).get_context_data(**kwargs)
+        context.update({'panels_with_dependencies':
```
Rewrap:

```
context.update({
    'panels_with_dependencies': {panel.get_name(): panel.dependencies for panel in self.request.user.get_all_panels()},
})
```

mitar commented on piplmesh/account/models.py:
```diff
@@ -89,11 +107,16 @@ class User(auth.User):
+        super(User, self).__init__(*args, **kwargs)
+        
+        # If the user has not previously saved any panel data, set up defaults
+        if not self.panels:
```
As I said, this does not seem to me as a good approach. Shouldn't you simply use `default` argument to field constructor?

Naltharial replied:
I tried, but that turned out to be the problem in that comment I said about `ValidationError`; `default` gets called before the panels are registered (in validation), so we can't get the list of panels then, only when the User object is actually being created.

mitar replied:
Which code exactly have you used?

mitar replied:
I just tried it and it works for me?

```
def default_panels():
    return {panel.get_name(): Panel() for panel in piplmesh_panels.panels_pool.get_all_panels()}
```

I removed `reset_panels` and `get_all_panels` methods and using `self.request.user.get_panels` instead of `self.request.user.get_all_panels` in view.

Naltharial replied:
I just put the dict constructor in the `default=`, but wrapping it in a function seems to work, since it probably calls it later. Where have you defined the function, though? If I put it in the class, I get `'default_panels' not defined` no matter the type of method.

mitar replied:
Use lambda functions.

mitar replied:
Read [here](https://docs.djangoproject.com/en/dev/ref/models/fields/#django.db.models.Field.default).

mitar replied:
And [here](https://mongoengine-odm.readthedocs.org/en/latest/guide/defining-documents.html#field-arguments).

mitar commented on piplmesh/account/models.py:
```diff
@@ -17,7 +17,7 @@
 
 from . import fields, utils
-from .. import panels
+from .. import panels as frontend_panels
```
`panels` module is on the same level as `frontend`. So better name could be `piplmesh_panels`.

mitar commented on piplmesh/account/templates/form/panels.html:
```diff
@@ -0,0 +1,28 @@
+
+{% load i18n frontend %}
+
+{% block form_field %}
```
Not correct. You have smaller blocks you can extend. You are still copy-pasting too much. What you need: you just need some additional information after each field, no? So which available block would be the most suitable for that?

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +206,71 @@ def create_user(cls, username, email=None, password=None):
+        layout.collapsed = collapsed
+        self.panels[name].layout[columns_count] = layout
+    
+    def get_columns(self, columns_count):
```
I must say that I don't understand this method. :-)

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +206,71 @@ def create_user(cls, username, email=None, password=None):
+        
+        if ordered:
+            if unordered_panels:
+                columns[-1] = unordered_panels
```
Is this `-1` supported by client code? What happens?

So if I understand correctly, the question is what to do with new panels which do not have layout. Here you are adding it to special column? Is this really necessary? What if we just omit those panels and leave to client-side code to put those panels somewhere?

So here is one question to be decided: should be panels without layout be put somewhere by server side or client side. If former, then you should put it somewhere real, some real column. If latter, you probably do not need to pass those panels through. :-)

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +206,71 @@ def create_user(cls, username, email=None, password=None):
+        """
+        
+        for panel, layout in self.get_layouts(columns_count).items():
+            layout.column, layout.order = column_ordering[panel]
```
Here is a slight question if `column_ordering[panel]` really exists. :-) You can add one if sentence around, to just skip the panel if `panel is not in column_ordering`.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +206,71 @@ def create_user(cls, username, email=None, password=None):
+        for name, panel in self.get_layouts(columns_count).items():
+            c = panel.column
+            
+            if c == None:
```
`if c is None`. `None` is always tested with `is`.

mitar commented on piplmesh/account/models.py:
```diff
@@ -187,3 +206,71 @@ def create_user(cls, username, email=None, password=None):
+                columns_order[c] = []
+
+            # Enforce ordering of panels in column, because layout is not ordered
+            pos = bisect.bisect_left(columns_order[c], panel.order)
```
Hm. Instead of using bisect, couldn't you simply sort data over which you are looping in correct order? And then you do not have to search and insert. In fact you are reimplementing insertion sort here. A bit.

```
for column, order, name in sorted([(panel.column, panel.order, name) for name, panel in self.get_layouts(columns_count).items() if panel.column is not None]):
    assert column >= 0
    while len(panels) <= column:
        panels.append([])
    panels[column].append(name)
```

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+$(document).ready(function () {
+    var panels_forcedby = {}
```
`;` at the end.

mitar replied:
`panels_forcedby` -> `panels_requiredby`.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+    $('#content .panels form').on('submit', function (event) {
+        $('input[type="checkbox"][disabled]', this).prop('disabled', false);
+    })
+    	.find('input[type="checkbox"]').on('change', function (event) {
```
Move this line to the end of previous line.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+    var panels_savestate = {}
+    
+    $('#content .panels form').on('submit', function (event) {
+        $('input[type="checkbox"][disabled]', this).prop('disabled', false);
```
`input:disabled[type="checkbox"]` maybe?

mitar replied:
Or `input[type="checkbox"]:disabled`?

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+    	.find('input[type="checkbox"]').on('change', function (event) {
+        validateCheckbox(this);
+    })
+    	.andSelf().find(':checked').change();
```
Move one line up.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+    
+    function validateCheckbox(checkbox) {
+        var checkbox = $(checkbox);
+        var panel = checkbox.attr('name');
```
Hm, I would use prop also here? No?

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+        var panel = checkbox.attr('name');
+        var checked = checkbox.prop('checked');
+        
+        if (panel in panels_with_dependencies && panels_with_dependencies[panel].length) {
```
Is this `if` really necessary? Is not this the same:

```
$.each(panels_with_dependencies[panel] || [], function (index, dependency) {
```

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+$(document).ready(function () {
+    var panels_forcedby = {}
+    var panels_savestate = {}
```
For this, you could use `data()` on DOM.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+        }
+        
+        // Trigger event to recursively check dependencies throughout hierarchy.
+        panel_checkbox.change()
```
`;` at the end.

Nice. :-) Is there a chance of infinite loop?

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+        var panel_locktext = $('.locked', panel_checkbox.parents('li'));
+        
+        if (!$.isEmptyObject(panels_forcedby[panel])) {
+            panel_locktext.text(gettext('Panel depended on by:'));
```
`"` for user exposed strings.

_Panel depended on by:_ -> _Panel is required by:_

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+        
+        if (!$.isEmptyObject(panels_forcedby[panel])) {
+            panel_locktext.text(gettext('Panel depended on by:'));
+            var list = $('<ul></ul>');
```
`$('<ul></ul>')` -> `$('<ul/>')`

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+            panel_locktext.text(gettext('Panel depended on by:'));
+            var list = $('<ul></ul>');
+            $.each(panels_forcedby[panel], function (panel, depends) {
+            	$('<li></li>').text(panel).appendTo(list);
```
Same.

Commit:
Code clarity.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Panels administration.

Naltharial commented:
I've replaced the `get_columns` method, the frontend does seem to work even with new panels that haven't been ordered yet. I had to use `panels_pool` in view though, because otherwise you don't get dependencies for panels that aren't currently selected.

Test cases also seem to pass now.

mitar commented on piplmesh/account/models.py:
```diff
@@ -53,6 +53,24 @@ class TwitterAccessToken(mongoengine.EmbeddedDocument):
+    """
+    This class holds panel instances for a user, their properties and layouts.
+    
+    :param dict layout: mapping of count of displayed columns (`string`) to `PanelState` objects for that count
```
Now I noticed. This is not really a param. You use this when commenting functions and which parameters they take. In this case this is a class. and even class constructor does not take this parameter.

You might simply move this line above `layout` definition bellow and just leave it as a in-code-only comment.

mitar commented on piplmesh/account/models.py:
```diff
@@ -89,11 +107,9 @@ class User(auth.User):
-    panels_order = mongoengine.DictField()
-
+    
+    panels = mongoengine.MapField(mongoengine.EmbeddedDocumentField(Panel), default=lambda: {panel.get_name(): Panel() for panel in piplmesh_panels.panels_pool.get_all_panels()})
```
Also comment here above, between what this maps.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,62 @@
+        var panel = checkbox.attr('name');
+        var checked = checkbox.prop('checked');
+        
+    	$.each(panels_with_dependencies[panel] || [], function (index, dependency) {
```
Indentation.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,62 @@
+            panel_locktext.text(gettext("Panel is required by:"));
+            var list = $('<ul/>');
+            $.each(panels_requiredby[panel], function (panel, depends) {
+            	$('<li/>').text(panel).appendTo(list);
```
Indentation.

mitar commented on piplmesh/account/templates/form/panels.html:
```diff
@@ -0,0 +1,23 @@
+{% block form_after_field %}
+    <div class="text">
+        <div class="locked">
+        </div>
```
Move this to the end of previous line.

mitar commented:
> the frontend does seem to work even with new panels that haven't been ordered yet.

Yes. This is good. So probably only frontend should decide how to order new panels. Not backend

> I had to use panels_pool in view though, because otherwise you don't get dependencies for panels that aren't currently selected.

I agree. But we had `get_all_panels` method or something on user object before, why is that one not good? It can return for now the same list as `panels_pool`. But later on it could be limited (by, for example, permissions (debug panel could be available only to developers), or payment (we could have some panels available only if you buy them)).

mitar commented:
Feature-wise, things are good now. So just check all your code for possible tabs and fix indentations and this is ready to go in.

There is one minor detail you can do, if you want, or open a ticket. You have to enable for elements before submit. But this is ugly as it changes user feedback and maybe confuses her. So some other solution should be used. What Internet recommends? Maybe a hidden input could be inserted which is kept in sync with checkbox? Maybe each checkbox could have its own hidden input all the time, kept in sync. JavaScript could rename all checkboxes and add hidden input box instead with that form name.

Commit:
Ghost panel inputs.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Unneccesary.

Naltharial commented:
I've added a second set of hidden inputs, so we don't have to un-disable the checkboxes. They're `IntegerField`, because hidden checkboxes don't properly submit state.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -165,3 +167,42 @@ def clean_confirmation_token(self):
+            data_box = panel.get_name()
+            display_box = data_box + "_display"
+            
+            attrs[data_box] = forms.IntegerField(required=False, initial=0)
```
Why not `BooleanField`, too?

mitar commented on piplmesh/account/forms.py:
```diff
@@ -165,3 +167,42 @@ def clean_confirmation_token(self):
+            display_box = data_box + "_display"
+            
+            attrs[data_box] = forms.IntegerField(required=False, initial=0)
+            attrs[data_box].widget = forms.HiddenInput()
```
You can pass this as an argument to the field above, no? Even just as a class?

`forms.IntegerField(required=False, initial=0, widget=forms.HiddenInput)`

This does not work?

Commit:
BooleanField, improved input detection.

Commit:
No debug.

Naltharial commented:
Fixed.

Commit:
No debug.

Commit:
Revert.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -165,3 +167,43 @@ def clean_confirmation_token(self):
+            attrs[display_box] = forms.BooleanField(label=data_box, required=False)
+            
+            # Connect display checkbox to data field
+            attrs[display_box].widget = forms.CheckboxInput(attrs={'data-panel':data_box, 'data-display':'True'})
```
Spaces after `:`.

mitar replied:
Why are you settings `widget` from outside and not through the constructor? Set it through the constructor, it is much cleaner (and it might be better handled, you never know if there is some special handling when passed through constructor which you are bypassing now).

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+$(document).ready(function () {
+    var panels_requiredby = {}
+    
+    $('#content .panels form input[type="checkbox"][data-display="True"]').on('change', function (event) {
```
I am not sure why is `data-display` needed? Are there any other checkboxes in the form?

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+$(document).ready(function () {
+    var panels_requiredby = {}
```
`;` missing at the end.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+    $('#content .panels form input[type="checkbox"][data-display="True"]').on('change', function (event) {
+        var display = $(this);
+        validateCheckbox(display);
+        $('#id_' + display.data('panel')).val(display.prop('checked'));
```
I don't like the idea that you assume here how ids are generated. You should retrieve this in some other manner.  Maybe store field if into data attribute, too?

mitar replied:
Do you really need data attribute? You do know that hidden and input fields are siblings in DOM (are they?) or at least that order of hidden inputs is the same as checkboxes.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -165,3 +167,43 @@ def clean_confirmation_token(self):
+class PanelFormMetaclass(forms.Form.__metaclass__):
+    def __new__(cls, name, bases, attrs):
+        for panel in panels.panels_pool.get_all_panels():
+            data_box = panel.get_name()
```
This string can (and should) be localized. You probably do not want to use it as form input name.

mitar replied:
I see that we have not defined localized names for panels. This is not good. This is bug.

OK, I see that we are everywhere using `get_name` as something machine-readable and constant. OK, leave `get_name` here, but we should define something like `verbose_name` or `display_name` then and use this everywhere where we show panel name to the user.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -165,3 +167,43 @@ def clean_confirmation_token(self):
+    def __new__(cls, name, bases, attrs):
+        for panel in panels.panels_pool.get_all_panels():
+            data_box = panel.get_name()
+            display_box = data_box + "_display"
```
Constant string should be in `'`. Use string interpolation instead of concatenation.

mitar commented on piplmesh/account/forms.py:
```diff
@@ -165,3 +167,43 @@ def clean_confirmation_token(self):
+            display_box = data_box + "_display"
+            
+            attrs[data_box] = forms.BooleanField(initial=False, widget=forms.HiddenInput(), required=False)
+            attrs[display_box] = forms.BooleanField(label=data_box, required=False)
```
But the label should be localized name of the panel.

mitar commented on piplmesh/account/static/piplmesh/js/panels.js:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    function lockCheckbox(panel) {
+        var panel_checkbox = $('#content .panels form input[data-panel="' + panel + '"]');
```
Hm. Do we really need data to access it through data? Each form element has its own unique form input name, at least (not to mention ID). Why not using that?

mitar commented:
You had some problems with link property? Or you just didn't upgrade from requirements.txt?

mitar commented:
I moved account Django app to separate Python package. You will have to update this pull request.

https://github.com/mitar/django-mongo-auth

mitar commented:
OK, now everything should work with external authentication package.
